snooty/parser.py

```diff
@@ -644,6 +650,99 @@ def handle_tabset(self, node: n.Directive) -> None:
             key=lambda x: tabid_list.index(cast(n.Directive, x).options["tabid"]),
         )
 
+    def handle_wayfinding(self, node: n.Directive) -> None:
+        expected_options = specparser.Spec.get().wayfinding["options"]
+        expected_options_dict = {option.id: option for option in expected_options}
+        expected_child_opt_name = "wayfinding-option"
+        expected_child_desc_name = "wayfinding-description"
+        expected_children_names = {expected_child_opt_name, expected_child_desc_name}
+        wayfinding_name = "wayfinding"
+
+        valid_opts: List[n.Node] = []
+        valid_desc: n.Node | None = None
+        used_ids: Set[str] = set()
+
+        # Validate children
+        for child in node.children:
+            child_line_start = child.start[0]
+
+            invalid_child = None
+            if not isinstance(child, n.Directive):
+                # Catches additional unwanted types like Paragraph
+                invalid_child = child.type
+            elif not child.name in expected_children_names:
+                invalid_child = child.name
+
+            if invalid_child:
+                self.diagnostics.append(
+                    InvalidChild(
+                        invalid_child,
+                        wayfinding_name,
+                        f"{expected_child_opt_name} or {expected_child_desc_name}",
+                        child_line_start,
+                    )
+                )
+                continue
+
+            # Type is ambiguous now despite if statements above
+            assert isinstance(child, n.Directive)
+            if child.name == expected_child_desc_name:
+                valid_desc = child
+                continue
+
+            option_id = child.options.get("id")
+
+            if not (child.argument and option_id):
+                # Don't append diagnostic since docutils should already
+                # complain about missing argument and ID option
+                continue
+
+            if not option_id in expected_options_dict:
+                self.diagnostics.append(
+                    UnknownWayfindingOption(option_id, child_line_start)
+                )
+                continue
+
+            if option_id in used_ids:
+                self.diagnostics.append(
+                    DuplicateWayfindingOption(option_id, child_line_start)
+                )
+                continue
+
+            option_details = expected_options_dict[option_id]
+            child.options["title"] = option_details.title
+            child.options["language"] = option_details.language
+            valid_opts.append(child)
+            used_ids.add(option_id)
+
+        def sort_key(node: n.Node) -> tuple[bool, str, str]:
```

Comment:
[minor] can we take the node arg to be type n.Directive

Reply:
Sure. I originally made it `n.Node` because `node.children` wasn't compatible with `List[n.Directive]`, which was causing typing issues with the list of nodes/directives here. Casting the list before reassignment seems to do the trick though. Let me know if it looks good to you